engine/new_runtime/src/memories/encrypted_ram.rs

```diff
@@ -0,0 +1,206 @@
+    ad: [u8; AD_SIZE],
+}
+
+// We currently only implement for u8 because our encryption functions only returns Vec[u8], therefore our cypher is Buffer<u8>
+impl<P: BoxProvider, const AD_SIZE: usize> LockedMemory<u8, P> for EncryptedRam<P, AD_SIZE> {
+    fn alloc(payload: &[u8], config: LockedConfiguration<P>) -> Result<Self, MemoryError> {
+        match config {
+            // For encrypted memory we don't store the key itself.
+            EncryptedRamConfig(key, size) => {
+                if size.is_none() {
+                    return Err(SizeNeededForAllocation);
+                }
+                let mut ad: [u8; AD_SIZE] = [0u8; AD_SIZE];
+                P::random_buf(&mut ad).or(Err(EncryptionError))?;
+                Ok(EncryptedRam {
+                    // Encrypt the payload before storing it
+                    cypher: {
+                        let encrypted_payload = P::box_seal(&key, &ad, payload).or(Err(EncryptionError))?;
+                        Buffer::alloc(&encrypted_payload, BufferConfig(encrypted_payload.len()))
+                            .expect("Failed to generate buffer")
+                    },
+                    // Don't put the actual key value, put random values, we don't want to store the key
+                    // for security reasons
+                    config: EncryptedRamConfig(Key::random(), size),
+                    ad,
+                })
+            }
+
+            // We don't allow any other configurations for EncryptedRam
+            _ => Err(ConfigurationNotAllowed),
+        }
+    }
+
+    /// Locks the memory and possibly reallocates
+    // Currently we reallocate a new EncryptedRam at each lock
+    // This improves security but decreases performance
+    fn lock(mut self, payload: Buffer<u8>, config: LockedConfiguration<P>) -> Result<Self, MemoryError> {
+        match config {
+            EncryptedRamConfig(_, _) => {
+                self.dealloc();
+                EncryptedRam::<P, AD_SIZE>::alloc(&payload.borrow(), config)
+            },
+            _ => Err(ConfigurationNotAllowed),
+        }
+    }
+
+    /// Unlocks the memory
+    fn unlock(&self, config: LockedConfiguration<P>) -> Result<Buffer<u8>, MemoryError> {
+        // assert_matches!(self.config, EncryptedRamConfig(_,_));
+
+        // Decrypt and store the value in a Buffer
+        if let EncryptedRamConfig(key, None) = config {
+            // Note: data is not in the protected buffer here, change box_open to return a Buffer type?
+            let data = P::box_open(&key, &self.ad, &*self.cypher.borrow()).or(Err(DecryptionError))?;
+            if let EncryptedRamConfig(_, Some(size)) = self.config {
+                Buffer::alloc(&data, BufferConfig(size))
+            } else {
+                panic!("This should not happen if EncryptedRam has been allocated properly")
+            }
+        } else {
+            Err(ConfigurationNotAllowed)
+        }
+    }
```

Comment:
I feel like the naming of the methods does not really match what they are doing. They actually don't change if/how the data is protected in memory. Instead `unlock` just simply borrows the data that is stored, and `lock` replaces the stored payload with a new payload. So if for example a user only reads a secret and never modifies it, it is not necessary to call `lock` (and unnecessarily re-allocate) because the payload didn't change. Same in `FileMemory`.

Reply:
Well that's a good point haha
Ok I will try to think of something else but feel free to give me some suggestions because my naming sense is terrible

Reply:
Well we could just go with simple `unlock` -> `borrow` and `write` -> `update`. Or we merge it to the one method `exec_on_unlock` that you listed as an idea, which takes a closure `Fn(Buffer<u8>) -> Option<Buffer<u8>>`. If the return value is `Some(data)` then we do the re-allocation, otherwise the stored payload does not change. But as you already mentioned as a _Con_ there is that it could get difficult if we have multiple steps of encryption.